### PR TITLE
Add DateTime constructor for Base.Libc.Tmstruct

### LINKED ISSUE #31524: ability to convert TmStruct to DateTime?
The "obvious" attempts to type-cast this didn't seem to work:
```julia
julia> DateTime(Libc.strptime("%Y-%M-%dT%H:%M:%SZ", "2018-11-16T10:26:14Z"))
ERROR: MethodError: no method matching Int64(::Base.Libc.TmStruct)
Closest candidates are:
  Int64(::Union{Bool, Int32, Int64, UInt32, UInt64, UInt8, Int128, Int16, Int8, UInt128, UInt16}) at boot.jl:707
  Int64(::Ptr) at boot.jl:717
  Int64(::Float32) at float.jl:700
  ...
Stacktrace:
 [1] DateTime(::Base.Libc.TmStruct, ::Int64, ::Int64, ::Int64, ::Int64, ::Int64, ::Int64) at /data/vtjnash/julia/usr/share/julia/stdlib/v1.2/Dates/src/types.jl:342 (repeats 2 times)
 [2] top-level scope at REPL[85]:1

julia> convert(DateTime, Libc.strptime("%Y-%M-%dT%H:%M:%SZ", "2018-11-16T10:26:14Z"))
ERROR: MethodError: Cannot `convert` an object of type Base.Libc.TmStruct to an object of type DateTime
Closest candidates are:
  convert(::Type{DateTime}, ::Date) at /data/vtjnash/julia/usr/share/julia/stdlib/v1.2/Dates/src/conversions.jl:30
  convert(::Type{DateTime}, ::Millisecond) at /data/vtjnash/julia/usr/share/julia/stdlib/v1.2/Dates/src/conversions.jl:34
  convert(::Type{T}, ::T) where T at essentials.jl:167
  ...
Stacktrace:
 [1] top-level scope at REPL[86]:1

julia> unix2datetime(time(Libc.strptime("%Y-%M-%dT%H:%M:%SZ", "2018-11-16T10:26:14Z")))
2018-01-16T15:26:14
```

### PR DESCRIPTION
**Pull request by eulerkochy**
Fixes #31524

### CODE REVIEW

**Comment by eulerkochy** [comment]
Could this be reviewed before feature freeze? Can this be considered as a `feature` @StefanKarpinski ?

**Comment by StefanKarpinski** [comment]
This seems sane to me. @quinnj: can you do a quick sanity check and merge if it's good?

File: stdlib/Dates/src/types.jl [hunk 1]
```diff
@@ -197,6 +197,8 @@ function validargs(::Type{DateTime}, y::Int64, m::Int64, d::Int64,
     return argerror()
 end
 
+DateTime(dt::Base.Libc.TmStruct) = Time(dt.hour, dt.min, dt.sec)
```
Comment (quinnj) [inline stdlib/Dates/src/types.jl]: Let's do this:
  * take a look at the `Libc.TmStruct` definition [here](https://github.com/JuliaLang/julia/blob/c5be8145fbd9b1f7acd68e81ca23a6a4976c46a9/base/libc.jl#L154)
  * for the `DateTime(dt::Base.Libc.TmStruct)` definition, let's make sure we also pull the year, month, day out of the `TmStruct` as well
  * add a separate `Time(dt::Base.Libc.TmStruct)` constructor that does what you have here

Sound good?

**Review by quinnj — CHANGES_REQUESTED** [review]

File: stdlib/Dates/src/types.jl (hunk 1 above)
Comment (quinnj) [reply stdlib/Dates/src/types.jl]: Part of the reasoning for my comment is to ensure that the `DateTime` constructor actually returns a `DateTime`. Having a separate `Time` constructor that returns a `Time` and does what you have here is also appropriate, or I guess people could just do `Time(DateTime(tmstruct))`

Comment (eulerkochy) [reply stdlib/Dates/src/types.jl]: @quinnj I have also added matching `Date` constructor, too!

**Comment by eulerkochy** [comment]
@quinnj is it alright now?

**Review by quinnj — CHANGES_REQUESTED** [review]
So here's what the docs say for the meaning of TmStruct fields:

Member | Type | Meaning | Range
-- | -- | -- | --
tm_sec | int | seconds after the minute | 0-61*
tm_min | int | minutes after the hour | 0-59
tm_hour | int | hours since midnight | 0-23
tm_mday | int | day of the month | 1-31
tm_mon | int | months since January | 0-11
tm_year | int | years since 1900 |  
tm_wday | int | days since Sunday | 0-6
tm_yday | int | days since January 1 | 0-365
tm_isdst | int | Daylight Saving Time flag

So we'll need to adjust the `DateTime` and `Date` constructors to account for the `TmStruct` month being 0-based (0 - 11), and the year being "years since 1900". So if we can make those changes and then add some tests making sure it's all being converted correctly, I think this is looking good.

**Comment by eulerkochy** [comment]
@quinnj thanks for the type definition, I've corrected the constructor definition, and have added tests.

**Comment by eulerkochy** [comment]
@StefanKarpinski @quinnj when is the feature freeze?

**Comment by StefanKarpinski** [comment]
Today.

**Review by quinnj — APPROVED** [review]
I think this looks good now!

**Comment by eulerkochy** [comment]
Bump!

**Comment by eulerkochy** [comment]
Bump

**Pull request merged**